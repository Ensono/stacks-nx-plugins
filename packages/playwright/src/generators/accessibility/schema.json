{
    "$schema": "http://json-schema.org/schema",
    "cli": "nx",
    "$id": "Playwright",
    "title": "Ensono Stacks Playwright Accessibility Generator",
    "description": "The @ensono-stacks/playwright:accessibility generator scaffolds your existing initialised Playwright project with example Accessibility tests and config.",
    "type": "object",
    "properties": {
        "project": {
            "type": "string",
            "description": "The name of the project.",
            "alias": "p",
            "$default": {
                "$source": "projectName"
            },
            "x-prompt": "What is the name of the project for this component?"
        }
    },
    "required": ["project"]
}
