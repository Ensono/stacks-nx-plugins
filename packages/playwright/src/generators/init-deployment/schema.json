{
  "$schema": "http://json-schema.org/schema",
  "$id": "InitDeployment",
  "title": "Ensono Stacks Playwright Init Deployment Generator",
  "description": "The @ensono-stacks/playwright:init-deployment generator adds the relevant pipeline and infrastructure code to your workspace.",
  "type": "object",
  "properties": {},
  "required": []
}
