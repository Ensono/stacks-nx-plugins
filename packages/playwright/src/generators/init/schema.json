{
  "$schema": "http://json-schema.org/schema",
  "$id": "Init",
  "title": "Ensono Stacks Playwright Init Generator",
  "description": "The @ensono-stacks/playwright:init generator initialises your playwright project with additional stacks configuration.",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "alias": "p",
      "$default": {
        "$source": "projectName"
      },
      "x-prompt": "What app would you like to generate a test project for?"
    }
  },
  "required": [
    "project"
  ]
}
