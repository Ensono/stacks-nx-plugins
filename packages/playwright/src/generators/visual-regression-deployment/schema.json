{
    "$schema": "http://json-schema.org/schema",
    "cli": "nx",
    "$id": "VisualRegressionDeployment",
    "title": "",
    "type": "object",
    "properties": {
        "type": {
            "type": "string",
            "description": "Method used to conduct visual regression testing",
            "alias": "t",
            "default": "none",
            "enum": ["native", "applitools"],
            "x-prompt": {
                "message": "What type of visual regression tests would you like to use?",
                "type": "list",
                "items": [
                    {
                        "value": "native",
                        "label": "Native"
                    },
                    {
                        "value": "applitools",
                        "label": "Applitools"
                    }
                ]
            }
        }
    },
    "required": ["type"]
}
