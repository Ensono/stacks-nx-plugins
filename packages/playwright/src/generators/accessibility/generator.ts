import { verifyPluginCanBeInstalled } from '@ensono-stacks/core';
import {
    addDependenciesToPackageJson,
    formatFiles,
    generateFiles,
    getProjects,
    offsetFromRoot,
    Tree,
} from '@nx/devkit';
import path from 'path';

import { AccessibilityGeneratorSchema } from './schema';
import {
    AXE_CORE_PLAYWRIGHT_VERSION,
    AXE_RESULTS_PRETTY_PRINT_VERSION,
} from '../../utils/versions';

interface NormalizedSchema extends AccessibilityGeneratorSchema {
    projectName: string;
    projectRoot: string;
}

function normalizeOptions(
    tree: Tree,
    options: AccessibilityGeneratorSchema,
): NormalizedSchema {
    const project = getProjects(tree).get(options.project);

    return {
        ...options,
        projectName: project?.name as string,
        projectRoot: project?.sourceRoot as string,
    };
}

function addFiles(tree: Tree, source: string, options: NormalizedSchema) {
    const templateOptions = {
        ...options,
        offsetFromRoot: offsetFromRoot(options.projectRoot),
        template: '',
    };

    generateFiles(
        tree,
        path.join(__dirname, source),
        options.projectRoot,
        templateOptions,
    );
}

async function updateDependencies(tree: Tree) {
    const accessibilityDeps = {
        '@axe-core/playwright': AXE_CORE_PLAYWRIGHT_VERSION,
        'axe-result-pretty-print': AXE_RESULTS_PRETTY_PRINT_VERSION,
    };

    return addDependenciesToPackageJson(tree, {}, accessibilityDeps);
}

export default async function accessibilityGenerator(
    tree: Tree,
    options: AccessibilityGeneratorSchema,
) {
    verifyPluginCanBeInstalled(tree, options.project);

    const project = getProjects(tree).get(options.project);

    if (!project) {
        throw new Error(`${options.project} does not exist.`);
    }

    if (!options.project.endsWith('-e2e')) {
        throw new Error(
            `${options.project} is not an e2e project. Please select a supported target.`,
        );
    }

    const normalizedOptions = normalizeOptions(tree, options);

    // generate acessiblity files
    addFiles(tree, 'files', normalizedOptions);

    await formatFiles(tree);

    return updateDependencies(tree);
}
