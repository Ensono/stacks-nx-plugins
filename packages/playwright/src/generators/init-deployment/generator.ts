import { formatFiles, Tree } from '@nrwl/devkit';

import { updateAzureDevopsStages } from './utils/update-azdevops-build';
import { updateTaskctlYaml, updateTasksYaml } from './utils/update-tasks-yamls';

export default async function initDeploymentGenerator(tree: Tree) {
    updateTaskctlYaml(tree);
    updateTasksYaml(tree);
    updateAzureDevopsStages(tree);

    await formatFiles(tree);
}
