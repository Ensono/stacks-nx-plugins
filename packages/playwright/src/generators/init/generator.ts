import {
    addIgnoreEntry,
    deploymentGeneratorMessage,
    execAsync,
    hasGeneratorExecutedForProject,
    tsMorphTree,
    verifyPluginCanBeInstalled,
} from '@ensono-stacks/core';
import initPlaywrightGenerator from '@mands/nx-playwright/src/generators/project/generator';
import { NxPlaywrightGeneratorSchema } from '@mands/nx-playwright/src/generators/project/schema-types';
import {
    formatFiles,
    generateFiles,
    getProjects,
    offsetFromRoot,
    Tree,
    readProjectConfiguration,
    addDependenciesToPackageJson,
    joinPathFragments,
    updateJson,
    runTasksInSerial,
} from '@nx/devkit';
import { Linter } from '@nx/eslint';
import path from 'path';

import { PlaywrightGeneratorSchema } from './schema';
import { updatePlaywrightConfigWithDefault } from './utils/update-playwright-config';
import { updatePlaywrightConfigBase } from './utils/update-playwright-config-base';
import { PLAYWRIGHT_VERSION } from '../../utils/versions';

interface NormalizedSchema extends PlaywrightGeneratorSchema {
    projectName: string;
    projectRoot: string;
}

function updateTsConfig(tree: Tree, project: string) {
    updateJson(tree, joinPathFragments(project, 'tsconfig.json'), tsconfig => ({
        ...tsconfig,
        compilerOptions: {
            ...tsconfig.compilerOptions,
            lib: ['dom'],
        },
    }));
}

function normalizeOptions(
    tree: Tree,
    options: PlaywrightGeneratorSchema,
): NormalizedSchema {
    const project = getProjects(tree).get(options.project);

    return {
        ...options,
        projectName: project?.name as string,
        projectRoot: project?.sourceRoot as string,
    };
}

function addFiles(tree: Tree, source: string, options: NormalizedSchema) {
    const templateOptions = {
        ...options,
        offsetFromRoot: offsetFromRoot(options.projectRoot),
        template: '',
    };

    const projectRootE2E = `${options.projectRoot}-e2e/src`;

    generateFiles(
        tree,
        path.join(__dirname, source),
        projectRootE2E,
        templateOptions,
    );
}

function updateDependencies(tree) {
    return addDependenciesToPackageJson(
        tree,
        {},
        {
            playwright: PLAYWRIGHT_VERSION,
            '@playwright/test': PLAYWRIGHT_VERSION,
            '@mands/nx-playwright': '^0.6.3',
        },
    );
}

export default async function initGenerator(
    tree: Tree,
    options: PlaywrightGeneratorSchema,
) {
    verifyPluginCanBeInstalled(tree, options.project);

    if (hasGeneratorExecutedForProject(tree, options.project, 'PlaywrightInit'))
        return false;

    const projectE2E = `${options.project}-e2e`;

    const optionsE2E = {
        ...options,
        projectE2E,
    };
    const normalizedOptionsForE2E = normalizeOptions(tree, optionsE2E);

    const playwrightGeneratorSchema: NxPlaywrightGeneratorSchema = {
        name: projectE2E,
        linter: Linter.EsLint,
        packageRunner: 'npx',
        project: options.project,
    };

    await initPlaywrightGenerator(tree, playwrightGeneratorSchema);

    const morphTree = tsMorphTree(tree);

    // playwright.config.base.ts
    updatePlaywrightConfigBase(morphTree);

    // add extra config to playwright.config.ts in project
    updatePlaywrightConfigWithDefault(
        readProjectConfiguration(tree, projectE2E),
        morphTree,
    );

    // example.spec.ts
    addFiles(tree, 'files', normalizedOptionsForE2E);

    // remove app.spec.ts added from @mands generato
    const { projectRoot } = normalizedOptionsForE2E;
    tree.delete(`${projectRoot}-e2e/src/app.spec.ts`);

    // Update tsconfig
    updateTsConfig(tree, `${projectRoot}-e2e`);

    // add records to gitignore
    addIgnoreEntry(tree, '.gitignore', 'Playwright', [
        '/test-results/',
        '/playwright-report/',
        '/playwright/.cache/',
    ]);

    await formatFiles(tree);

    return runTasksInSerial(
        updateDependencies(tree),
        () =>
            deploymentGeneratorMessage(
                tree,
                'nx g @ensono-stacks/playwright:init-deployment',
            ),
        () => execAsync('npx playwright install', projectRoot) as Promise<void>,
    );
}
