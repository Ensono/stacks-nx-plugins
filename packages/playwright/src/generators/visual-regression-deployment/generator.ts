import {
    hasGeneratorExecutedForWorkspace,
    verifyPluginCanBeInstalled,
} from '@ensono-stacks/core';
import { formatFiles, Tree, logger } from '@nx/devkit';

import { VisualRegressionDeploymentGeneratorSchema } from './schema';
import { updateAzureDevopsStagesApplitools } from './utils/update-azdevops-stage';
import { updateAzureDevopsSnapshotsYaml } from './utils/update-azure-devops-updatesnapshots';
import { updateTaskctlYaml, updateTasksYaml } from './utils/update-tasks-yamls';
import { visualRegressionTypes } from '../../utils/types';

// eslint-disable-next-line consistent-return
export default async function visualRegressionDeploymentGenerator(
    tree: Tree,
    options: VisualRegressionDeploymentGeneratorSchema,
) {
    verifyPluginCanBeInstalled(tree);

    if (
        hasGeneratorExecutedForWorkspace(
            tree,
            'PlaywrightVisualRegressionDeployment',
            true,
        )
    )
        return false;

    switch (options.type) {
        case visualRegressionTypes.NATIVE: {
            // update tasks.yaml
            updateTasksYaml(tree);
            // update taskctl.yaml
            updateTaskctlYaml(tree, { visualRegression: true });

            updateAzureDevopsSnapshotsYaml(tree);
            break;
        }
        case visualRegressionTypes.APPLITOOLS: {
            updateAzureDevopsStagesApplitools(tree);

            logger.warn(`Don't forget to set your 'APPLITOOLS_API_KEY'.`);
            break;
        }
        default: {
            await formatFiles(tree);
        }
    }
}
