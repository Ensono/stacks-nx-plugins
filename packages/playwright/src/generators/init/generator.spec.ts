import { Tree, readProjectConfiguration } from '@nrwl/devkit';
import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';

import generator from './generator';
import { PlaywrightGeneratorSchema } from './schema';

describe('playwright generator', () => {
    let appTree: Tree;
    const options: PlaywrightGeneratorSchema = {
        project: 'test',
        accessibility: false,
        visualRegression: 'none',
    };

    beforeEach(() => {
        appTree = createTreeWithEmptyWorkspace();
    });

    it('should run successfully', async () => {
        await generator(appTree, options);
        const config = readProjectConfiguration(appTree, 'test');
        expect(config).toBeDefined();
    });
});
