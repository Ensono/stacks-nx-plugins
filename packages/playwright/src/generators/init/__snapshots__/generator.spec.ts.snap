// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`playwright generator should run successfully with default options: .prettierrc 1`] = `
"{ "singleQuote": true }
"
`;

exports[`playwright generator should run successfully with default options: apps/test-e2e/playwright.config.ts 1`] = `
"import { defineConfig, devices } from '@playwright/test';
import { nxE2EPreset } from '@nx/playwright/preset';

import { workspaceRoot } from '@nx/devkit';

// For CI, you may want to set BASE_URL to the deployed application.
const baseURL = process.env['BASE_URL'] || 'http://127.0.0.1:3000';

/**
 * Read environment variables from file.
 * https://github.com/motdotla/dotenv
 */
// require('dotenv').config();

/**
 * See https://playwright.dev/docs/test-configuration.
 */
export default defineConfig({
  ...nxE2EPreset(__filename, { testDir: './src' }),
  /* Shared settings for all the projects below. See https://playwright.dev/docs/api/class-testoptions. */
  use: {
    baseURL,
    /* Collect trace when retrying the failed test. See https://playwright.dev/docs/trace-viewer */
    trace: 'on-first-retry',
  },
  /* Run your local dev server before starting the tests */
  webServer: {
    command: 'npx nx start test',
    url: 'http://127.0.0.1:3000',
    reuseExistingServer: !process.env.CI,
    cwd: workspaceRoot,
  },
  projects: [
    {
      name: 'chromium',
      use: { ...devices['Desktop Chrome'] },
    },

    {
      name: 'firefox',
      use: { ...devices['Desktop Firefox'] },
    },

    {
      name: 'webkit',
      use: { ...devices['Desktop Safari'] },
    },

    // Uncomment for mobile browsers support
    /* {
      name: 'Mobile Chrome',
      use: { ...devices['Pixel 5'] },
    },
    {
      name: 'Mobile Safari',
      use: { ...devices['iPhone 12'] },
    }, */

    // Uncomment for branded browsers
    /* {
      name: 'Microsoft Edge',
      use: { ...devices['Desktop Edge'], channel: 'msedge' },
    },
    {
      name: 'Google Chrome',
      use: { ...devices['Desktop Chrome'], channel: 'chrome' },
    } */
  ],
});
"
`;

exports[`playwright generator should run successfully with default options: apps/test-e2e/src/example.spec.ts 1`] = `
"import { test, expect } from '@playwright/test';

test('has title', async ({ page }) => {
  await page.goto('/');

  // Expect h1 to contain a substring.
  expect(await page.locator('h1').innerText()).toContain('Welcome');
});
"
`;

exports[`playwright generator should run successfully with default options: nx.json 1`] = `
"{
  "affected": {
    "defaultBase": "main"
  },
  "targetDefaults": {
    "build": {
      "cache": true
    },
    "lint": {
      "cache": true
    },
    "@nx/js:tsc": {
      "cache": true,
      "dependsOn": ["^build"],
      "inputs": ["default", "^default"]
    },
    "@nx/eslint:lint": {
      "cache": true,
      "inputs": [
        "default",
        "{workspaceRoot}/.eslintrc.json",
        "{workspaceRoot}/.eslintignore",
        "{workspaceRoot}/eslint.config.js"
      ]
    },
    "@nx/jest:jest": {
      "cache": true,
      "inputs": ["default", "^default", "{workspaceRoot}/jest.preset.js"],
      "options": {
        "passWithNoTests": true
      },
      "configurations": {
        "ci": {
          "ci": true,
          "codeCoverage": true
        }
      }
    }
  },
  "stacks": {
    "config": {
      "business": {
        "company": "Amido",
        "domain": "stacks",
        "component": "nx"
      },
      "domain": {
        "internal": "test.com",
        "external": "test.dev"
      },
      "cloud": {
        "region": "euw",
        "platform": "azure"
      },
      "pipeline": "azdo",
      "terraform": {
        "group": "terraform-group",
        "storage": "terraform-storage",
        "container": "terraform-container"
      },
      "vcs": {
        "type": "github",
        "url": "remote.git"
      }
    },
    "executedGenerators": {
      "project": {
        "test": ["PlaywrightInit"]
      },
      "workspace": []
    }
  }
}
"
`;

exports[`playwright generator should run successfully with default options: package.json 1`] = `
"{
  "name": "@proj/source",
  "dependencies": {
    "tslib": "^2.3.0"
  },
  "devDependencies": {
    "@nx/devkit": "18.3.4",
    "@nx/eslint": "18.3.4",
    "@nx/eslint-plugin": "18.3.4",
    "@nx/jest": "18.3.4",
    "@nx/js": "18.3.4",
    "@nx/playwright": "18.3.4",
    "@playwright/test": "^1.44.0",
    "@swc-node/register": "~1.8.0",
    "@swc/core": "~1.3.85",
    "@swc/helpers": "~0.5.2",
    "@types/jest": "^29.4.0",
    "@types/node": "18.16.9",
    "@typescript-eslint/eslint-plugin": "^7.3.0",
    "@typescript-eslint/parser": "^7.3.0",
    "eslint": "~8.57.0",
    "eslint-config-prettier": "^9.0.0",
    "eslint-plugin-playwright": "^0.15.3",
    "jest": "^29.4.1",
    "jest-environment-jsdom": "^29.4.1",
    "playwright": "^1.44.0",
    "prettier": "^2.6.2",
    "ts-jest": "^29.1.0",
    "ts-node": "10.9.1",
    "typescript": "~5.4.2"
  }
}
"
`;

exports[`playwright generator should run successfully with default options: tsconfig.base.json 1`] = `
"{
  "compilerOptions": {
    "paths": {
      "@proj/test-e2e": ["apps/test-e2e/src/index.ts"]
    }
  }
}
"
`;
